clients/src/test/java/org/apache/kafka/common/log/remote/storage/LocalTieredStorageEvent.java

```diff
@@ -0,0 +1,147 @@
+package org.apache.kafka.common.log.remote.storage;
+
+import org.apache.kafka.common.*;
+
+import java.util.*;
+
+import static java.lang.String.*;
+import static java.util.Objects.*;
+import static java.util.Optional.*;
+
+public final class LocalTieredStorageEvent implements Comparable<LocalTieredStorageEvent> {
+
+    public enum EventType {
+        OFFLOAD_SEGMENT,
+        FETCH_SEGMENT,
+        FETCH_OFFSET_INDEX,
+        FETCH_TIME_INDEX,
+        DELETE_SEGMENT
+    }
+
+    private final EventType type;
+    private final int brokerId;
+    private final RemoteLogSegmentId segmentId;
+    private final int timestamp;
+    private final Optional<RemoteLogSegmentFileset> fileset;
+    private final Optional<RemoteLogSegmentMetadata> metadata;
+    private final Optional<Long> startPosition;
+    private final Optional<Long> endPosition;
+    private final Optional<Exception> exception;
+
+    public boolean matches(final LocalTieredStorageCondition condition) {
+        if (condition.eventType != type) {
+            return false;
+        }
+        if (condition.brokerId != brokerId) {
+            return false;
+        }
+        if (!segmentId.topicPartition().equals(condition.topicPartition)) {
+            return false;
+        }
+        if (!exception.map(e -> condition.failed).orElseGet(() -> !condition.failed)) {
+            return false;
+        }
+        return true;
+    }
+
+    public EventType getType() {
+        return type;
+    }
+
+    public TopicPartition getTopicPartition() {
+        return segmentId.topicPartition();
+    }
+
+    public boolean isAfter(final LocalTieredStorageEvent event) {
+        return event.timestamp < timestamp;
+    }
+
+    @Override
+    public int hashCode() {
+        return timestamp % 31;
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        return (other instanceof LocalTieredStorageEvent)
+                && (((LocalTieredStorageEvent)other).timestamp == timestamp);
```

Comment:
This is because the logical timestamp assigned to an event is designed to be a unique id for this event, that is: only one event corresponds to a given id, and a given id can only be assigned to one event.
_Timestamp_ here does not refer to the value read from a clock but a number in a strictly monotonic sequence. I guess it would be beneficial to update the field name.
Event is actually very close to an identity object. To diverge from the identity equality, an event has to be artificially constructed, that is, created outside of its functional scope.

Reply:
Correction - actually, the sequence is relative to an instance of `LocalTieredStorage` so it does not work.
I removed the override for `hashCode` and `equals` for now and moved back to identity equal.